docs/rfc/rfc-105-event-observer.md

```diff
@@ -0,0 +1,240 @@
+places where events are published via the event bus, e.g.:
+
+- <https://github.com/cometbft/cometbft/blob/091a1f312e5f2f4b183fab1d57d729a6c478ff1f/consensus/state.go#L743>
+- <https://github.com/cometbft/cometbft/blob/091a1f312e5f2f4b183fab1d57d729a6c478ff1f/consensus/state.go#L964>
+- <https://github.com/cometbft/cometbft/blob/091a1f312e5f2f4b183fab1d57d729a6c478ff1f/consensus/state.go#L971>
+- etc.
+
+All of these event publishing methods, not only for consensus state but for
+other types of event publishers, are defined on this central `EventBus` type,
+which seems to signal that its functionality should be decentralized.
+
+### Strongly Typed Event Observer
+
+A simple alternative pattern here would be to define an **event observer**
+interface for every major component of the system capable of producing events.
+For consensus state, this may end up looking something like:
+
+```golang
+package consensus
+
+// StateObserver is specific to the consensus.State struct, and all of its
+// methods are called from within consensus.State instead of using an "event
+// bus". This allows for greater compile-time guarantees through doing away with
+// the generic pub/sub mechanism in the event bus.
+//
+// Note how all methods are infallible (i.e. they do not return any errors).
+// This is functionally equivalent to the fire-and-forget pattern implemented by
+// the event bus.
+//
+// Also note how method names are prefixed by the name of the relevant producer
+// of events (in this case "ConsensusState", corresponding to the
+// consensus.State struct). This is intentional to allow composition of
+// observers of multiple different components without function names clashing.
+//
+// Finally, given that this is just straightforward Go, it is up to either the
+// caller or the callee to decide how to handle the concurrency of certain
+// events. The event bus approach, by contrast, is always concurrent and relies
+// on Go channels, which could end up filling up and causing back-pressure into
+// the caller (already observed in slow WebSocket subscribers).
+type StateObserver interface {
+    ConsensusStateNewRoundStep(ev EventDataRoundState)
+    ConsensusStateTimeoutPropose(ev EventDataRoundState)
+    ConsensusStateTimeoutWait(ev EventDataRoundState)
+    // ...
+}
+```
+
+And on `consensus.State` one could easily either supply such an observer in the
+constructor, or define a new method that allows one to set the observer, e.g.:
+
+```golang
+package consensus
+
+func (cs *State) SetObserver(obs StateObserver) {
+    cs.observer = obs
+}
+```
+
+Then, instead of publishing events via the event bus, one would simply do the
+following:
+
+```diff
+-        if err := cs.eventBus.PublishEventNewRoundStep(rs); err != nil {
+-            cs.Logger.Error("failed publishing new round step", "err", err)
+-        }
++        // Notify the observer
++        cs.observer.ConsensusStateNewRoundStep(rs)
+```
+
```

Comment:
I'm not sure where, but I would add somewhere in the text that this pattern will also be able to do away with cumbersome interfaces like [SwitchToConsensus](https://github.com/cometbft/cometbft/blob/3c88a5ec/blocksync/reactor.go#L35) and its awkward (and ugly) [callsite](https://github.com/cometbft/cometbft/blob/3c88a5ec/blocksync/reactor.go#L383-L385)

https://github.com/cometbft/cometbft/blob/3c88a5ec0b12f28c1646268ba39950f2a7cc4013/blocksync/reactor.go#L385